BlazarService/src/main/java/com/hubspot/blazar/listener/SlackNotificationVisitor.java

```diff
@@ -0,0 +1,188 @@
+  private static final Optional<String> ABSENT_STRING = Optional.absent();
+
+  private SlackConfigurationService slackConfigurationService;
+  private final BranchService branchService;
+  private final ModuleBuildService moduleBuildService;
+  private final BlazarUrlHelper blazarUrlHelper;
+  private final SlackClient slackClient;
+  private final RepositoryBuildService repositoryBuildService;
+
+  @Inject
+  public SlackNotificationVisitor(SlackConfigurationService slackConfigurationService,
+                                  BranchService branchService,
+                                  ModuleBuildService moduleBuildService,
+                                  BlazarUrlHelper blazarUrlHelper,
+                                  SlackClient slackClient,
+                                  RepositoryBuildService repositoryBuildService) {
+    this.slackConfigurationService = slackConfigurationService;
+    this.branchService = branchService;
+    this.moduleBuildService = moduleBuildService;
+    this.blazarUrlHelper = blazarUrlHelper;
+    this.slackClient = slackClient;
+    this.repositoryBuildService = repositoryBuildService;
+  }
+
+
+  @Override
+  public void visit(RepositoryBuild build) throws Exception {
+    Set<SlackConfiguration> configurationSet = slackConfigurationService.getAllWithBranchId(build.getBranchId());
+    for (SlackConfiguration slackConfiguration : configurationSet) {
+      Optional<RepositoryBuild> previous = repositoryBuildService.getPreviousBuild(build);
+      if (previous.isPresent() && !shouldSend(slackConfiguration, build.getState(), previous.get())) {
+        continue;
+      }
+      GitInfo gitInfo = branchService.get(build.getBranchId()).get();
+      sendSlackMessage(slackConfiguration, build, gitInfo);
+    }
+  }
+
+  private boolean shouldSend(SlackConfiguration slackConfiguration, RepositoryBuild.State state, RepositoryBuild previous) {
+    // OnChange
+    boolean changedState = previous.getState() == state;
+    if (slackConfiguration.getOnChange() && changedState) {
+      return true;
+    }
+    // OnFinish
+    if (slackConfiguration.getOnFinish() && state.isComplete()) {
+      return true;
+    }
+    // OnRecovery
+    boolean previousFailed = previous.getState().isComplete() && previous.getState() == RepositoryBuild.State.SUCCEEDED;
+    if (slackConfiguration.getOnRecover() && previousFailed && state == RepositoryBuild.State.SUCCEEDED) {
+      return true;
+    }
+    // OnFailure
+    if (slackConfiguration.getOnFail() && FAILED_REPO_STATES.contains(state)) {
+      return true;
+    }
+    return false;
+  }
+
+  private void sendSlackMessage(SlackConfiguration slackConfiguration, RepositoryBuild build, GitInfo gitInfo) {
+    String fallback = String.format("Module Build %s-%s-%s finished with state %s", gitInfo.getRepository(), gitInfo.getRepository(), gitInfo.getBranch(), build.getState().toString());
+    Optional<String> color = Optional.absent();
```

Comment:
you have `ABSENT_STRING` defined as a constant and used below - might as well use it here too?